File: 2022/04.hs
module Four () where

import Utils (atoiList)

parse :: String -> [Int]
parse = atoiList . map (\c -> if c == '-' then ',' else c)

contains [a1, a2, b1, b2] = if (a1 <= b1 && b2 <= a2) || (b1 <= a1 && a2 <= b2) then 1 else 0

overlaps [a1, a2, b1, b2] = if (a1 <= b1 && a2 >= b1) || (b1 <= a1 && b2 >= a1) then 1 else 0

partOne = sum . map contains
partTwo = sum . map overlaps

main :: IO ()
main = do
  contents <- getContents
  let input = lines contents
      theLines = map parse input

  print "partOne"
  print $ partOne theLines

  print "partTwo"
  print $ partTwo theLines

File: 2022/04.js
const R = require('ramda')
const readline = require('readline')
const rl = readline.createInterface({ input: process.stdin })

const lines = []

rl.on('line', data => {
  const line = data.replaceAll('-', ',').split(',').map(i => parseInt(i))

  lines.push(line)
})

rl.on('close', () => {
  console.log('partOne', partOne(lines))
  console.log('partTwo', partTwo(lines))
})

function contains ([a1, a2, b1, b2]) {
  if (
    (a1 <= b1 && b2 <= a2) ||
    (b1 <= a1 && a2 <= b2)
  ) return 1

  return 0
}

function partOne (lines) {
  return R.sum(R.map(contains, lines))
}

function partTwo (lines) {
  return 'todo'
}
